chainerx_cc/chainerx/cuda/cuda_device/linalg.cu

```diff
@@ -0,0 +1,120 @@
+#include <cuda_runtime.h>
+#include <cuda_fp16.hpp>
+
+#include "chainerx/array.h"
+#include "chainerx/axes.h"
+#include "chainerx/backend.h"
+#include "chainerx/backend_util.h"
+#include "chainerx/cuda/cublas.h"
+#include "chainerx/cuda/cusolver.h"
+#include "chainerx/cuda/cuda_runtime.h"
+#include "chainerx/cuda/cuda_set_device_scope.h"
+#include "chainerx/cuda/data_type.cuh"
+#include "chainerx/cuda/float16.cuh"
+#include "chainerx/cuda/kernel_regist.h"
+#include "chainerx/device.h"
+#include "chainerx/dtype.h"
+#include "chainerx/error.h"
+#include "chainerx/float16.h"
+#include "chainerx/kernels/creation.h"
+#include "chainerx/kernels/linalg.h"
+#include "chainerx/kernels/math.h"
+#include "chainerx/kernels/misc.h"
+#include "chainerx/macro.h"
+#include "chainerx/routines/creation.h"
+#include "chainerx/routines/math.h"
+
+namespace chainerx {
+namespace cuda {
+
+class CudaCholeskyKernel : public CholeskyKernel {
+public:
+    void Call(const Array& a, const Array& out) override {
+        Device& device = a.device();
+        device.CheckDevicesCompatible(a, out);
+        CudaSetDeviceScope scope{device.index()};
+
+        if (a.ndim() != 2 || out.ndim() != 2) {
+            throw DimensionError{"ChainerX cholesky supports only 2-dimensional arrays."};
+        }
+        if (a.shape()[0] != a.shape()[1]) {
+            throw DimensionError{"Matrix is not square."};
+        }
+
+        // potrf (cholesky) stores result in-place, therefore copy ``a`` to ``out`` and then pass ``out`` to the routine
+        device.backend().CallKernel<CopyKernel>(a, out);
+
+        bool is_out_contiguous = out.IsContiguous();
+        Array out_contiguous = is_out_contiguous ? out : AsContiguous(out);
+
+        auto cholesky_impl = [&](auto pt, auto bufsize_func, auto solver_func) {
+            CHAINERX_ASSERT(a.dtype() == out_contiguous.dtype());
+
+            using T = typename decltype(pt)::type;
+
+            // Note that cuSOLVER uses Fortran order.
+            // To compute a lower triangular matrix L = cholesky(A), we use cuSOLVER to compute an upper triangular matrix U = cholesky(A).
+            cublasFillMode_t uplo = CUBLAS_FILL_MODE_UPPER;
+
+            cuda_internal::DeviceInternals& device_internals = cuda_internal::GetDeviceInternals(static_cast<CudaDevice&>(device));
+
+            // compute workspace size and prepare workspace
+            T* out_ptr = static_cast<T*>(internal::GetRawOffsetData(out_contiguous));
+            int work_size = 0;
+            const int N = a.shape()[0];
+            device_internals.cusolver_handle().Call(
+                bufsize_func,
+                uplo,
+                N,
+                out_ptr,
+                N,
+                &work_size);
+
+            // POTRF execution
+            T* work_space;
+            CheckCudaError(cudaMalloc(&work_space, work_size * sizeof(T)));
+            int *devInfo;
+            CheckCudaError(cudaMalloc(&devInfo, sizeof(int)));
```

Comment:
Could you use the `Allocate` function of the CUDA device wrapper similar to how we allocate workspace size in `chainerx/cuda/cuda_conv.cc` to avoid calling these CUDA functions directly.

Reply:
I'll change it to use `Allocate` function.
I am just more familiar with the pure CUDA functions :slightly_smiling_face:

Reply:
Workspace is now allocated using `Empty` array.
I don't know what would be a better way than current for the `devInfo`. It needs to be in GPU memory and then copied to CPU memory to check for errors.

Reply:
`devInfo` is now allocated using `device.Allocate` and then copied to CPU with `device.MemoryCopyTo`.